### ISSUE TITLE
Windows MSVC building failed with c++20

### ISSUE THREAD

**Issue by LiMYmy** [Simple-Robotics/proxsuite#154]
It works with c++17 but has error with c++20 on Windows MSVC(```VisualStudio 2022 17.4.3```). But will work with Clang15, both c++17 and c++20.
This is the CMakeLists.txt file.
```
cmake_minimum_required(VERSION 3.14)

project(proxqp_test)
set(CMAKE_VERBOSE_MAKEFILE ON)

include(FetchContent)
# set(FETCHCONTENT_UPDATES_DISCONNECTED ON)

FetchContent_Declare(
    simde
    GIT_REPOSITORY https://github.com/simd-everywhere/simde.git
    GIT_TAG v0.7.2
    GIT_SHALLOW    TRUE
)
FetchContent_MakeAvailable(simde)
set(Simde_INCLUDE_DIR ${simde_SOURCE_DIR})

set(BUILD_TESTING OFF CACHE INTERNAL "")
FetchContent_Declare(
    proxsuite
    GIT_REPOSITORY https://github.com/Simple-Robotics/proxsuite.git
    GIT_TAG v0.2.16
    GIT_SHALLOW    TRUE
)
FetchContent_MakeAvailable(proxsuite)

# find_package(proxsuite REQUIRED)

add_executable(app main.cpp)
target_link_libraries(app PUBLIC
  proxsuite
)
target_compile_features(app PUBLIC cxx_std_20)
target_compile_definitions(app PUBLIC NOMINMAX)
target_compile_options(app PUBLIC "/arch:AVX2")
```

**Comment by jcarpent** [comment]
Could you provide the log report for C++20 issues on MSVC?
We often suffers from issues with MSVC which does not implement standards like others ...

**Comment by LiMYmy** [comment]
@jcarpent Will Chinese be ok? I don't know how to change the output language.
There are so many error messages, here is the first few lines.
```
[build]   cl /c /I"E:\mycode\proxqp\build\_deps\proxsuite-src\include" /I"E:\mycode\proxqp\build\_deps\proxsuite-build\include" /Zi /W3 /WX- /diagnostics:column /O2 /Ob1 /D _MBCS /D WIN32 /D _WINDOWS /D NDEBUG /D NOMINMAX /D "CMAKE_INTDIR=\"RelWithDebInfo\"" /Gm- /EHsc /MD /GS /fp:precise /Zc:wchar_t /Zc:forScope /Zc:inline /GR /std:c++20 /permissive- /Fo"app.dir\RelWithDebInfo\\" /Fd"app.dir\RelWithDebInfo\vc143.pdb" /external:W0 /Gd /TP /errorReport:queue  /external:I "D:/Anaconda3/Library/include/eigen3" /bigobj E:\mycode\proxqp\main.cpp
[build] E:\mycode\proxqp\build\_deps\proxsuite-src\include\proxsuite/linalg/veg/internal/macros.hpp(1151,43): error C2397: 从“size_t”转换到“proxsuite::linalg::veg::isize”需要收缩转换 [E:\mycode\proxqp\build\app.vcxproj]
[build] E:\mycode\proxqp\build\_deps\proxsuite-src\include\proxsuite/linalg/veg/ref.hpp(122,1): error C2061: 语法错误: 标识符“__VEG_PP_TAIL” [E:\mycode\proxqp\build\app.vcxproj]
[build] E:\mycode\proxqp\build\_deps\proxsuite-src\include\proxsuite/linalg/veg/ref.hpp(123,72): error C3861: “r”: 找不到标识符 [E:\mycode\proxqp\build\app.vcxproj]
[build] E:\mycode\proxqp\build\_deps\proxsuite-src\include\proxsuite/linalg/veg/ref.hpp(118,3): message : 此诊断出现在编译器生成的函数“proxsuite::linalg::veg::RefMut<_detail::_meta::uncvlref<T&>::type> proxsuite::linalg::veg::nb::mut::operator ()(void) noexcept const”中 [E:\mycode\proxqp\build\app.vcxproj]
[build] E:\mycode\proxqp\build\_deps\proxsuite-src\include\proxsuite/linalg/veg/ref.hpp(123,72): error C2065: “r”: 未声明的标识符 [E:\mycode\proxqp\build\app.vcxproj]
[build] E:\mycode\proxqp\build\_deps\proxsuite-src\include\proxsuite/linalg/veg/ref.hpp(118,3): message : 此诊断出现在编译器生成的函数“proxsuite::linalg::veg::RefMut<_detail::_meta::uncvlref<T&>::type> proxsuite::linalg::veg::nb::mut::operator ()(void) noexcept const”中 [E:\mycode\proxqp\build\app.vcxproj]
[build] E:\mycode\proxqp\build\_deps\proxsuite-src\include\proxsuite/linalg/veg/ref.hpp(131,1): error C2061: 语法错误: 标识符“__VEG_PP_TAIL” [E:\mycode\proxqp\build\app.vcxproj]
[build] E:\mycode\proxqp\build\_deps\proxsuite-src\include\proxsuite/linalg/veg/ref.hpp(134,14): error C3861: “arg”: 找不到标识符 [E:\mycode\proxqp\build\app.vcxproj]
[build] E:\mycode\proxqp\build\_deps\proxsuite-src\include\proxsuite/linalg/veg/ref.hpp(127,3): message : 此诊断出现在编译器生成的函数“T proxsuite::linalg::veg::nb::clone::operator ()(void) noexcept(<expr>) const”中 [E:\mycode\proxqp\build\app.vcxproj]
[build] E:\mycode\proxqp\build\_deps\proxsuite-src\include\proxsuite/linalg/veg/ref.hpp(134,14): error C2065: “arg”: 未声明的标识符 [E:\mycode\proxqp\build\app.vcxproj]
[build] E:\mycode\proxqp\build\_deps\proxsuite-src\include\proxsuite/linalg/veg/ref.hpp(127,3): message : 此诊断出现在编译器生成的函数“T proxsuite::linalg::veg::nb::clone::operator ()(void) noexcept(<expr>) const”中 [E:\mycode\proxqp\build\app.vcxproj]
```
This is the first error at ```proxsuite-src/include/proxsuite/linalg/veg/internal/macros.hpp```. The red underline is below ```isize```.
```
template<usize... Is, StrLiteralImpl<isize{ sizeof...(Is) }> L>
struct StrLiteralExpand<_meta::integer_sequence<usize, Is...>, L>
{
  using Type = StrLiteralConstant<L._[Is]...>;
};
```

**Comment by jcarpent** [comment]
Could you provide the English version instead? You should be able to find the trick on msvc doc.

**Comment by LiMYmy** [comment]
I couldn't make it in vscode, so I changed to visual studio and got these messages.
<img width="757" alt="image" src="https://user-images.githubusercontent.com/72388584/209352877-4103e756-b3e8-40da-bc01-73b59f3df493.png">

**Comment by jcarpent** [comment]
@aescande Do you have any idea how to fix this issue on Windows?
Thanks in advanced for your help and guidance.

**Comment by LiMYmy** [comment]
I found two errors. The first one is about a conversion from ```unsigned long long``` to ```signed long long``` and the second is the usage of c++20 ```concepts```. 
I just enforce the conversion and ```VEG_HAS_CONCEPTS``` to make it work.[717eff1](https://github.com/LiMYmy/proxsuite/commit/717eff1c882a936b5b57db65b3c042fb70cb4c1f)

**Comment by jcarpent** [comment]
Solved via #155.
